### PR TITLE
feat(toolkit): improve docker build time in CI

### LINKED ISSUE #1748: Improve docker build time in CI environment
Hello,

In a CI environment where no layer cache are available, docker builds of ECS tasks can take a very long time.

A classical solution for this is to pull first the latest image from the registry and use it as an input for the `--cache-from` argument of `docker build` (see here https://github.com/aws/aws-codebuild-docker-images/issues/26#issuecomment-370177343 and https://medium.com/@gajus/making-docker-in-docker-builds-x2-faster-using-docker-cache-from-option-c01febd8ef84).

This implies the following changes in [`prepareContainerAsset`](https://github.com/awslabs/aws-cdk/blob/master/packages/aws-cdk/lib/docker.ts#L22):
* In addition to tagging the image with the `calculateImageFingerprint`, tag with `latest`
* When in CI (detected by something like `process.env.CI`) try to pull `latest` first and then build with `--cache-from`

([`toolkitInfo.prepareEcrRepository`](https://github.com/awslabs/aws-cdk/blob/63ae0b4b5ebea077010940fb52e951c9664a5261/packages/aws-cdk/lib/api/toolkit-info.ts#L102) should also be refactored/splitted in order to get the `repostoryUri` sooner/without having to pass a `tag`)

I would be interested to have aws-cdk's team opinion on this.

### PR DESCRIPTION
**Pull request by jogold**
When running in CI, try to pull the latest image first and use it as cache for the build. CI is detected by the presence of the `CI` environment variable or the `--ci` command line option.

When pushing docker images, always tag with `latest`.

Closes #1748

To be discussed: name of env var for detection, `CI` is set by lots of CI providers (not by CodeBuild though), use a opt-in mechanism instead?

Test (with empty ECR):
```sh
$ docker images -a | grep "/cdk/" | awk '{print $3}' | xargs docker rmi # or a more specific grep
$ cdk deploy # should build and push image
$ docker images -a | grep "/cdk/" | awk '{print $3}' | xargs docker rmi --force # or a more specific grep
$ CI=true cdk deploy # should pull, build almost instantly but not push
```
----

### Pull Request Checklist

* [ ] Testing
  - Unit test added
  - __CLI change?:__ coordinate update of integration tests with team
  - __cdk-init template change?:__ coordinated update of integration tests with team
* [ ] Docs
  - __jsdocs__: All public APIs documented
  - __README__: README and/or documentation topic updated
* [x] Title and Description
  - __Change type__: title prefixed with **fix**, **feat** will appear in changelog
  - __Title__: use lower-case and doesn't end with a period
  - __Breaking?__: last paragraph: "BREAKING CHANGE: <describe what changed + link for details>"
  - __Issues__: Indicate issues fixed via: "**Fixes #xxx**" or "**Closes #xxx**"
* [ ] Sensitive Modules (requires 2 PR approvers)
  - IAM Policy Document (in @aws-cdk/aws-iam)
  - EC2 Security Groups and ACLs (in @aws-cdk/aws-ec2)
  - Grant APIs (only if not based on official documentation with a reference)

----

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license.

### CODE REVIEW

File: packages/aws-cdk/lib/docker.ts [hunk 1]
```diff
@@ -95,6 +132,9 @@ async function calculateImageFingerprint(imageId: string) {
   // Metadata that has no bearing on the image contents
   delete manifest.Created;
 
+  // Parent can change when using --cache-from in CI
+  delete manifest.Parent;
```
Comment (rix0rrr) [inline packages/aws-cdk/lib/docker.ts]: What does `Parent` represent here? Please tell me it has nothing to do with disk images

File: packages/aws-cdk/lib/docker.ts [hunk 2]
```diff
@@ -18,43 +18,69 @@ import { PleaseHold } from './util/please-hold';
...
-        '--password', ecr.password,
-        ecr.endpoint]);
+      if (!loggedIn) { // We could be already logged in if in CI
+        await dockerLogin(toolkitInfo);
+      }
 
       const qualifiedImageName = `${ecr.repositoryUri}:${tag}`;
+
       await shell(['docker', 'tag', imageId, qualifiedImageName]);
+      await shell(['docker', 'tag', imageId, latest]); // Tag with `latest` also
```
Comment (rix0rrr) [inline packages/aws-cdk/lib/docker.ts]: You need to describe this change in the PR description as well.

File: packages/aws-cdk/lib/docker.ts (hunk 1 above)
Comment (jogold) [reply packages/aws-cdk/lib/docker.ts]: It's the sha256 of the docker's parent image. When using `--cache-from`, the sha256 of the parent gets modified, even for the same final image.

**Comment by jogold** [comment]
> Can you describe a little more about the rationale behind the change? Are we not trading `docker build` time for `docker pull` time? Are we guaranteed that pulling will be faster than building?

We are indeed trading `docker build` time for `docker pull` time. Generally pulling will be faster than building for Dockerfiles with OS packages installation and code packages installation (pip, npm, etc.). If nothing has changed in the current commit (in CI), it will be faster to pull than to build. If there are changes, it's more likely to see changes in the code that is copied in the image (COPY/ADD, the bottom layers) and again pulling should be faster than building.

For very simple Dockerfiles (FROM + COPY/ADD, few layers) or for a complete change of the Dockerfile, I can imagine that it could be faster to build.

> Why limit this behavior change to CI builds only?

When running locally chances are that we already have a lot of layer cache available. In this case, pulling will almost always be slower than building.

> Though I like the use of the `CI` environment variable, I think there needs to be a command-line switch `--ci` which is threaded through to the build step. When not supplied, it could default to the presence of the environment variable `CI`. Be aware that `--no-ci` should also work to disable an autodetected CI build (it'll be the difference between `undefined` and `false` in the yargs object).

OK, I will look into this.

**Comment by rix0rrr** [comment]
Thanks for the explanation. I'd like to see that rationale captured in one or two sentenced in a code comment somewhere.

File: packages/aws-cdk/lib/docker.ts (hunk 1 above)
Comment (rix0rrr) [reply packages/aws-cdk/lib/docker.ts]: Wait, if it's a parent image, and the parent image changes on every build, then we cannot assume that the "current" image is the same either, can we?

Comment (jogold) [reply packages/aws-cdk/lib/docker.ts]: In this case the sha256 of the layers will change normally. I added this when I saw that for the same digest (on ECR) when using `--cache-from`, `calculateImageFingerprint` returned a different finger print, resulting in 3 tags on the same image (latest, fingerprint1 and fingerprint2). This needs further investigation. The optimal solution may be a mix between the custom finger print calculation and the docker digest. Looking into this.

Comment (jogold) [reply packages/aws-cdk/lib/docker.ts]: I'm removing this `delete manifest.Parent`, it's a mistake and not needed here. The sha256 of the Parent is also present in the Config.Image key so deleting it in one place only should not have had an effect.

File: packages/aws-cdk/bin/cdk.ts [hunk 3]
```diff
@@ -61,6 +61,7 @@ async function parseCommandLineArguments() {
     .command('deploy [STACKS..]', 'Deploys the stack(s) named STACKS into your AWS account', yargs => yargs
       .option('exclusively', { type: 'boolean', alias: 'e', desc: 'only deploy requested stacks, don\'t include dependencies' })
       .option('require-approval', { type: 'string', choices: [RequireApproval.Never, RequireApproval.AnyChange, RequireApproval.Broadening], desc: 'what security-sensitive changes need manual approval' }))
+      .option('ci', { type: 'boolean', desc: 'Force CI detection. Use --no-ci to disable CI autodetection.', default: undefined })
```
Comment (jogold) [inline packages/aws-cdk/bin/cdk.ts]: @rix0rrr The description makes no reference to docker/asset preparation. I thought that this option could be used elsewhere in the future. What about the documentation for this new option? I see that https://github.com/awsdocs/aws-cdk-user-guide/blob/master/doc_source/tools.md is not up to date...

Comment (rix0rrr) [reply packages/aws-cdk/bin/cdk.ts]: That seems good. @Doug-AWS will update the tools section once we release.

Please follow this by something like:

```
if (argv.ci === undefined) {
   argv.ci = process.env.CI !== undefined;
}
```

To make sure that afterwards `argv.ci` is well-defined. Saves having to redo the CI-autodetection logic in every place where it's used. Configuring:

```
default: process.env.CI !== undefined
```

To force the default value to either `true` or `false` might also work.

File: packages/aws-cdk/lib/docker.ts [hunk 4]
```diff
@@ -18,38 +18,71 @@ import { PleaseHold } from './util/please-hold';
...
 
   const buildHold = new PleaseHold(` ⌛ Building Docker image for ${asset.path}; this may take a while.`);
   try {
+    const ecr = await toolkitInfo.prepareEcrRepository(asset.id);
+    const latest = `${ecr.repositoryUri}:latest`;
+
+    let loggedIn = false;
+
+    // In CI we try to pull latest first
+    if (ci === true || (process.env.CI && ci !== false)) {
```
Comment (rix0rrr) [inline packages/aws-cdk/lib/docker.ts]: Don't do the CI autodetection here, move it up to `cdk.ts`.

File: packages/aws-cdk/lib/docker.ts [hunk 5]
```diff
@@ -18,38 +18,71 @@ import { PleaseHold } from './util/please-hold';
...
+    }
+
     buildHold.start();
 
-    const command = ['docker',
+    const baseCommand = ['docker',
       'build',
       '--quiet',
       asset.path];
+    const command = process.env.CI
```
Comment (rix0rrr) [inline packages/aws-cdk/lib/docker.ts]: Need to use the variable here.

File: packages/aws-cdk/lib/docker.ts (hunk 4 above)
Comment (Doug-AWS) [reply packages/aws-cdk/lib/docker.ts]: Yep. I have a script that recreates the 'cdk --help' part of the tools section whenever we have a new release.

File: packages/aws-cdk/lib/assets.ts [hunk 6]
```diff
@@ -10,7 +10,7 @@ import { zipDirectory } from './archive';
 import { prepareContainerAsset } from './docker';
 import { debug, success } from './logging';
 
-export async function prepareAssets(stack: SynthesizedStack, toolkitInfo?: ToolkitInfo, ci?: boolean): Promise<CloudFormation.Parameter[]> {
+export async function prepareAssets(stack: SynthesizedStack, ci: boolean, toolkitInfo?: ToolkitInfo): Promise<CloudFormation.Parameter[]> {
```
Comment (jogold) [inline packages/aws-cdk/lib/assets.ts]: Had to reorder `ci` and `toolkitInfo` here (_a required parameter cannot follow an optional parameter_)

File: packages/aws-cdk/bin/cdk.ts (hunk 3 above)
Comment (Doug-AWS) [reply packages/aws-cdk/bin/cdk.ts]: Tools is updated. I should have a new public build of the guide by EOW.

**Pull request merged**